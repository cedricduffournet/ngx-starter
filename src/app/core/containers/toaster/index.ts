export * from './toaster.component';
export * from './toaster.model';
