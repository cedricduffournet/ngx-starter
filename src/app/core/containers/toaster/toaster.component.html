<toaster-container [toasterconfig]="toasterConfig"></toaster-container>
