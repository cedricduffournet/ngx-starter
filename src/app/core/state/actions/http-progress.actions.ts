import { createAction, props } from '@ngrx/store';

export const begin = createAction(
  '[HTTP progress] Begin http call',
  props<{ url: string }>()
);

export const complete = createAction(
  '[HTTP complete] Complete http call',
  props<{ url: string }>()
);
