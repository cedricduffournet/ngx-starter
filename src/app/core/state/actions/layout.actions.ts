import { createAction } from '@ngrx/store';

export const toggleMenu = createAction('[Layout Core] toggle menu');
