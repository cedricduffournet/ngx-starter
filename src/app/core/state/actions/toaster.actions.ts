import { props, createAction } from '@ngrx/store';
import { ToasterParams } from '@app/core/containers/toaster';
export const pop = createAction(
  '[Toaster] Pop toaster',
  props<{ params: ToasterParams }>()
);
