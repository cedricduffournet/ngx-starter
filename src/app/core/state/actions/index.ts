import * as LayoutCoreActions from './layout.actions';
import * as ToasterActions from './toaster.actions';
import * as HttpProgressActions from './http-progress.actions';

export { LayoutCoreActions, ToasterActions, HttpProgressActions };
