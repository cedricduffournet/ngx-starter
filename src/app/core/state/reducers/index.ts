import { InjectionToken } from '@angular/core';
import {
  createFeatureSelector,
  createSelector,
  ActionReducerMap,
  Action
} from '@ngrx/store';
import * as fromRouter from '@ngrx/router-store';
import * as fromLayout from '@app/core/state/reducers/layout.reducer';
import * as fromToaster from '@app/core/state/reducers/toaster.reducer';
import * as fromHttpProgress from '@app/core/state/reducers/http-progress.reducer';
import * as fromAuth from '@app/authentication/state/reducers';
import { RouterStateUrl } from '@app/core/utils/custom-route-serializer';

export interface State {
  layout: fromLayout.State;
  toaster: fromToaster.State;
  http: fromHttpProgress.State;
  router: fromRouter.RouterReducerState<RouterStateUrl>;
}

export const reducers = new InjectionToken<ActionReducerMap<State, Action>>(
  'Root reducers token',
  {
    factory: () => ({
      layout: fromLayout.reducer,
      toaster: fromToaster.reducer,
      http: fromHttpProgress.reducer,
      router: fromRouter.routerReducer
    })
  }
);

export const getRouterState = createFeatureSelector<
  fromRouter.RouterReducerState<RouterStateUrl>
>('router');

export const getPageTitle = createSelector(
  getRouterState,
  router => {
    if (router.state.data && router.state.data.title) {
      return router.state.data.title;
    }
    return '';
  }
);

export const getPreviousLink = createSelector(
  getRouterState,
  state => {
    if (state.state.data && state.state.data.previousLink) {
      return state.state.data.previousLink;
    }
    return null;
  }
);

export const getLayoutState = createFeatureSelector<State, fromLayout.State>(
  'layout'
);

export const getShowSidenav = createSelector(
  getLayoutState,
  fromLayout.getShowSidenav
);

export const getMenu = createSelector(
  getLayoutState,
  fromLayout.getMenu
);

export const getUserMenu = createSelector(
  getMenu,
  fromAuth.getLoggedUser,
  (menu, loggedUser) => {
    if (!loggedUser) {
      return [];
    }
    return menu.filter(item =>
      item.roles.some(role => loggedUser.roles.indexOf(role) !== -1)
    );
  }
);

export const getToasterState = createFeatureSelector<State, fromToaster.State>(
  'toaster'
);

export const getToasterParams = createSelector(
  getToasterState,
  fromToaster.getParams
);

export const getHttpState = createFeatureSelector<
  State,
  fromHttpProgress.State
>('http');

export const getHttpProcessList = createSelector(
  getHttpState,
  fromHttpProgress.getProcessList
);

export const isHttpProcessing = createSelector(
  getHttpProcessList,
  processList => processList.length !== 0
);
