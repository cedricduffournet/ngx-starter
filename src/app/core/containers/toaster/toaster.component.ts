import { Component, OnInit } from '@angular/core';
import { Store, select } from '@ngrx/store';

import { Observable } from 'rxjs';
import { skip } from 'rxjs/operators';
import {
  ToasterConfig,
  ToasterService
} from 'angular2-toaster/angular2-toaster';

import {
  defaultConfig,
  ToasterParams
} from '@app/core/containers/toaster/toaster.model';
import * as fromRoot from '@app/core/state/reducers';

@Component({
  selector: 'app-toaster',
  templateUrl: './toaster.component.html'
})
export class ToasterComponent implements OnInit {
  public toasterConfig: ToasterConfig = new ToasterConfig(defaultConfig);

  public toasterParams$: Observable<ToasterParams>;

  public constructor(
    private store: Store<fromRoot.State>,
    public toaster: ToasterService
  ) {}

  ngOnInit() {
    this.toasterParams$ = this.store.pipe(select(fromRoot.getToasterParams));

    this.toasterParams$.pipe(skip(1)).subscribe((params: ToasterParams) => {
      if (params.timeout && params.timeout > 0) {
        this.toasterConfig.timeout = params.timeout;
      }
      if (params.clear) {
        this.toaster.clear();
      }
      this.toaster.pop(params.type, params.title, params.body);
    });
  }
}
