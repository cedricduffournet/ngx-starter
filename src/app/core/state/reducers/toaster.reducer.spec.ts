import { reducer } from '@app/core/state/reducers/toaster.reducer';
import * as fromToaster from '@app/core/state/reducers/toaster.reducer';

import { ToasterActions } from '@app/core/state/actions';

describe('ToasterReducer', () => {
  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      expect(result).toMatchSnapshot();
    });
  });

  describe('POP', () => {
    it('should set params', () => {
      const params = {
        type: 'success',
        title: 'title',
        body: 'body',
        clear: true,
        timeout: 12
      };

      const action = ToasterActions.pop({
        params
      });

      const result = reducer(fromToaster.INITIAL_STATE, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('getParams', () => {
      it('shoud retrieve params', () => {
        const result = fromToaster.getParams({
          ...fromToaster.INITIAL_STATE,
          params: {
            type: 'success',
            title: 'title',
            body: 'body',
            clear: true,
            timeout: 12
          }
        });

        expect(result).toMatchSnapshot();
      });
    });
  });
});
