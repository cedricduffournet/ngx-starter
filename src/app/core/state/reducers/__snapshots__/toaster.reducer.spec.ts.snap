// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToasterReducer POP should set params 1`] = `
Object {
  "params": Object {
    "body": "body",
    "clear": true,
    "timeout": 12,
    "title": "title",
    "type": "success",
  },
}
`;

exports[`ToasterReducer Selectors getParams shoud retrieve params 1`] = `
Object {
  "body": "body",
  "clear": true,
  "timeout": 12,
  "title": "title",
  "type": "success",
}
`;

exports[`ToasterReducer undefined action should return the default state 1`] = `
Object {
  "params": Object {
    "body": "",
    "clear": false,
    "timeout": null,
    "title": "",
    "type": "",
  },
}
`;
